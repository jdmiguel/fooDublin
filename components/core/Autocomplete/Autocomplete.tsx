import { useState, useEffect, useRef } from 'react';
import { Loader } from '../Loader/Loader';
import { Input } from '../Input/Input';
import { Card } from '../Card/Card';
import { BlockText } from '../BlockText/BlockText';
import { Button } from '../Button/Button';
import {
  StyledAutocomplete,
  StyledInputWrapper,
  StyledListboxWrapper,
  StyledLoaderWrapper,
  StyledListbox,
  StyledNoSuggestionsWrapper,
  StyledErrorWrapper,
  StyledErrorButtonWrapper,
  StyledListboxItem,
} from './styles/autocomplete';
import { PlaceholderText, DEFAULT_TEXT_LOADING } from '@/store/statics';
import { Restaurant } from '../../pages/types';
import { CardType } from '../types';

export type AutocompleteProps = {
  hasSearchIcon?: boolean;
  suggestions: Restaurant[];
  loading: boolean;
  className?: string;
  disabled: boolean;
  onRequestError: boolean;
  fetchSuggestions: (search: string) => void;
  selectSuggestion: (id: number, name: string) => void;
  clearSuggestions: () => void;
};

export const Autocomplete: React.FC<AutocompleteProps> = ({
  suggestions,
  fetchSuggestions,
  selectSuggestion,
  loading,
  hasSearchIcon,
  className,
  disabled,
  onRequestError,
  clearSuggestions,
}) => {
  const blurDelay = useRef(0);
  const [isSuggestable, setIsSuggestable] = useState(true);
  const [value, setValue] = useState('');
  const [focusedPlaceholder, setFocusedPlaceholder] = useState(PlaceholderText.BLURRED);
  const [isListboxFocused, setIsListboxFocused] = useState(false);

  useEffect(() => {
    if (value.length > 2 && isSuggestable) {
      fetchSuggestions(value);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value, isSuggestable]);

  const hasBorderBottomRadius = !isListboxFocused || (isListboxFocused && value.length < 3);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const currentValue = event.target.value;

    setIsListboxFocused(currentValue.length > 2);
    setValue(currentValue);
  };

  const handleInputFocus = () => {
    if (!value) {
      setFocusedPlaceholder(PlaceholderText.FOCUSED);
    } else if (value.length > 2) {
      setIsListboxFocused(true);
    }
  };

  const handleInputBlur = () => {
    !value && setFocusedPlaceholder(PlaceholderText.BLURRED);
    blurDelay.current = window.setTimeout(() => {
      setIsListboxFocused(false);
    }, 100);
  };

  const handleSuggestionClick = (restaurantId: number, showedText: string) => {
    clearTimeout(blurDelay.current);

    setValue(showedText);
    setIsListboxFocused(false);
    setIsSuggestable(false);

    selectSuggestion(restaurantId, showedText);
  };

  const handleClearSuggestion = () => {
    setValue('');
    clearSuggestions();
  };

  const renderSuggestions = () => {
    if (suggestions.length === 0 && value.length > 2) {
      return (
        <StyledNoSuggestionsWrapper>
          <BlockText text="There are no suggestions for this search" />
        </StyledNoSuggestionsWrapper>
      );
    }

    return suggestions.map(({ id, imgSrc, title, content }: Restaurant) => (
      <StyledListboxItem key={id} role="option">
        <Card
          imgSrc={imgSrc}
          title={title}
          content={content}
          onClick={() => handleSuggestionClick(id, title)}
          type={CardType.SUGGESTION}
        />
      </StyledListboxItem>
    ));
  };

  return (
    <StyledAutocomplete data-testid="autocomplete" className={className} disabled={disabled}>
      <StyledInputWrapper hasBorderBottomRadius={hasBorderBottomRadius}>
        <Input
          type="text"
          placeholder={focusedPlaceholder}
          onChange={handleChange}
          value={value}
          onFocus={handleInputFocus}
          onBlur={handleInputBlur}
          hasSearchIcon={hasSearchIcon}
        />
      </StyledInputWrapper>
      <StyledListboxWrapper isShown={isListboxFocused} data-testid="listbox-wrapper">
        {loading ? (
          <StyledLoaderWrapper>
            <Loader className="listbox-loader" text={DEFAULT_TEXT_LOADING} />
          </StyledLoaderWrapper>
        ) : (
          <StyledListbox role="listbox">
            {onRequestError ? (
              <StyledErrorWrapper>
                <BlockText text="Sorry but something was wrong..." />
                <StyledErrorButtonWrapper>
                  <Button onClick={handleClearSuggestion}>Try again</Button>
                </StyledErrorButtonWrapper>
              </StyledErrorWrapper>
            ) : (
              renderSuggestions()
            )}
          </StyledListbox>
        )}
      </StyledListboxWrapper>
    </StyledAutocomplete>
  );
};
