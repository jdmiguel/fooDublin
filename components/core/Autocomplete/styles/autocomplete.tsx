import styled from 'styled-components';

export const StyledAutocomplete = styled.div<{ disabled: boolean }>`
  width: 100%;
  max-width: 550px;
  height: 55px;
  position: relative;
  cursor: pointer;
  pointer-events: ${({ disabled }) => disabled && 'none'};
`;

export const StyledInputWrapper = styled.div<{
  hasBorderBottomRadius: boolean;
}>`
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  border-radius: 4px;
  transition: background-color 0.2s ease-out;
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }

  ${({ hasBorderBottomRadius }) =>
    !hasBorderBottomRadius &&
    `border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;`}
`;

export const StyledListboxWrapper = styled.div<{ isShowed: boolean }>`
  visibility: ${({ isShowed }) => (isShowed ? 'visible' : 'hidden')};
  opacity: ${({ isShowed }) => (isShowed ? '1' : '0')};
  transform: translateY(${({ isShowed }) => (isShowed ? '0' : '10px')});
  transition: opacity 0.2s ease 0s, transform 0.2s ease 0s;
  position: absolute;
  box-sizing: border-box;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  z-index: 2;
  top: 55px;
  left: 0;
  padding: 0;
  width: 100%;
  height: auto;
  max-height: 400px;
  overflow: auto;
  box-shadow: 0 2px 2px rgba(0, 0, 0, 0.35);
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  border-radius: 0 0 4px 4px;
  border-bottom-right-radius: 4px;
  border-bottom-left-radius: 4px;
  border-top: 0;
`;

export const StyledLoaderWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

export const StyledListbox = styled.ul`
  width: 100%;
  -webkit-tap-highlight-color: transparent;
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

export const StyledNoSuggestionsWrapper = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

export const StyledErrorWrapper = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

export const StyledErrorButtonWrapper = styled.div`
  width: 150px;
  margin-top: 20px;
`;

export const StyledListboxItem = styled.li`
  width: 100%;
  border-bottom: 1px solid ${({ theme }) => theme.palette.LIGHT_MIN};
  list-style: none;
  display: flex;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  transition: background-color 0.2s ease-out;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
`;
