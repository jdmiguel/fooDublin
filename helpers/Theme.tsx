import React from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

export const theme: DefaultTheme = {
  palette: {
    PRIMARY: '#EC7E00',
    PRIMARY_DARK: '#D48C24',
    PRIMARY_MEDIUM: '#FBA52B',
    PRIMARY_LIGHT: '#FDF3E5',
    SECONDARY: '#660000',
    LIGHT_MAX: '#FFFFFF',
    LIGHT_MEDIUM: '#F3F3F3',
    LIGHT_MIN: '#E5EAED',
    LIGHT_SOFT: '#D8D8D8',
    DARK_MAX: '#222C3C',
    DARK_MEDIUM: '#585858',
    DARK_MIN: '#89959B',
    DARK_SOFT: '#A7A7A7',
  },
};

export const renderWithTheme = (component: any) => (
  <ThemeProvider theme={theme}>{component}</ThemeProvider>
);
