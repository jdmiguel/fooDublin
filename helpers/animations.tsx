import { keyframes, css } from 'styled-components';

const fadeIn = (endValue: number) => keyframes`
  0% { opacity: 0 }
  100% { opacity: ${endValue} }
`;

const fadeInFromPosY = (initialPosY: number) => keyframes`
  0% { opacity: 0; transform: translateY(${initialPosY}px); }
  100% { opacity: 1; ; transform: translateY(0); }
`;

const rotate = keyframes`
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
`;

const moveX = keyframes`
  0% { transform: translateX(-100%); }
  100% { transform: translateX(0); }
`;

export const fadeInAnimation = css<{ endValue?: number }>`
  animation: ${({ endValue }) => fadeIn(endValue || 1)} 0.15s linear;
`;

export const fadeInFromPosYAnimation = css<{
  initialPosY?: number;
  animationDuration?: number;
  animationDelay?: number;
}>`
  animation-name: ${({ initialPosY }) => fadeInFromPosY(initialPosY || 50)};
  animation-duration: ${({ animationDuration }) => animationDuration || 0.15}s;
  animation-delay: ${({ animationDelay }) => animationDelay || 0.25}s;
  animation-fill-mode: forwards;
  animation-timing-function: ease-out;
`;

export const circleLoaderAnimation = css`
  animation: ${rotate} 0.8s linear infinite;
`;

export const lineLoaderAnimation = css`
  animation: ${moveX} 1s ease-in-out infinite forwards;
`;
