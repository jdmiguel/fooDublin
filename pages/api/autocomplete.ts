import axios from 'axios';
import { NextApiRequest, NextApiResponse } from 'next';
import { BASE_API } from '@/store/statics';

export default async (req: NextApiRequest, res: NextApiResponse) => {
  const { searchText, latitude, longitude } = req.query;

  try {
    const response = await axios(`${BASE_API}autocomplete`, {
      method: 'GET',
      headers: {
        'Access-Control-Allow-Origin': '*',
        'x-requested-with': 'xmlhttprequest',
        accept: 'application/json',
        Authorization: `Bearer ${process.env.YELP_API_KEY}`,
      },
      params: {
        text: searchText,
        latitude,
        longitude,
      },
    });

    res.status(200).json(response.data.businesses);
  } catch (error: any) {
    res.status(error.response?.status || 500).json({ error: 'An error occurred' });
  }
};
