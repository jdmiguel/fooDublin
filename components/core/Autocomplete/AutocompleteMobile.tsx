import React, { useState, useEffect, useRef } from 'react';
import Link from 'next/link';
import styled from 'styled-components';
import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';

import { Input } from '../Input/Input';
import { Card }, { CardType } from '../Card/Card';
import { Loader } from '../Loader/Loader';
import { BlockTitle } from '../BlockTitle/BlockTitle';

import {
  PlaceholderText,
  DEFAULT_TEXT_LOADING,
} from '../../../helpers/staticData';
import { Restaurant } from '../../../helpers/types';

export type AutocompleteMobileProps = {
  className?: string;
  hasSearchIcon?: boolean;
  suggestions: Restaurant[];
  loading: boolean;
  fetchSuggestions: (search: string) => void;
  selectSuggestion: (id: string, name: string) => void;
};

const StyledAutocompleteMobile = styled.div`
  width: 100%;
  max-width: 550px;
  height: 100%;
`;

const StyledLabel = styled.div`
  position: relative;
  border-radius: 4px;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
`;

const StyledLabelButton = styled.button`
  width: 100%;
  height: 55px;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  line-height: 55px;
  padding: 0 4px 0 8px;
  display: flex;
  align-items: center;
  border-radius: 4px;
  cursor: pointer;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  i {
    font-size: 1.2rem;
    margin-left: 5px;
    margin-right: 10px;
    color: ${({ theme }) => theme.palette.DARK_SOFT};
  }
  span {
    color: ${({ theme }) => theme.palette.DARK_SOFT};
  }
`;

const StyledModal = styled.div<{ isShowed: boolean }>`
  visibility: ${({ isShowed }) => (isShowed ? 'visible' : 'hidden')};
  opacity: ${({ isShowed }) => (isShowed ? '1' : '0')};
  transition: opacity 0.2s ease 0s;
  position: fixed;
  box-sizing: border-box;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  z-index: 2;
  max-width: 767px;
  top: 0;
  left: 0;
  padding: 20px 20px 10px;
  width: 100%;
  height: 100%;
  box-shadow: 0 2px 2px rgba(0, 0, 0, 0.35);
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  border-radius: 0 0 4px 4px;
  border-bottom-right-radius: 4px;
  border-bottom-left-radius: 4px;
  border-top: 0;
`;

const StyleHeading = styled.div`
  margin: 15px 0 25px;
  display: flex;
  justify-content: space-between;
`;

const StyleHeadingText = styled(BlockTitle)`
  font-size: 1.3rem;
`;

const StyleHeadingButton = styled.button`
  display: block;
  transition: opacity 0.2s ease-out;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  i {
    font-size: 1.5rem;
    font-weight: 600;
    color: ${({ theme }) => theme.palette.DARK_MEDIUM};
  }
`;

const StyledInput = styled(Input)`
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_SOFT};
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  transition: background-color 0.2s ease-out;
  outline: none;
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
  &:focus {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }

  input {
    cursor: pointer;
  }
`;

const StyledListboxWrapper = styled.div<{ isShowed: boolean }>`
  visibility: ${({ isShowed }) => (isShowed ? 'visible' : 'hidden')};
  opacity: ${({ isShowed }) => (isShowed ? '1' : '0')};
  transition: opacity 0.2s ease 0s;
  position: absolute;
  box-sizing: border-box;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  z-index: 1;
  top: 150px;
  left: 0;
  padding: 0 10px 173px;
  margin-top: 15px;
  width: 100%;
  height: 100%;
  max-height: 100vh;
  overflow: auto;
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  border-radius: 4px;
`;

const StyledLoader = styled(Loader)`
  width: 100%;
  height: 100%;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

const StyledListbox = styled.ul`
  width: 100%;
`;

const StyledListboxItem = styled.li`
  width: 100%;
  border-bottom: 1px solid ${({ theme }) => theme.palette.LIGHT_MIN};
  list-style: none;
  overflow: hidden;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  display: flex;
  transition: background-color 0.2s ease-out;
  outline: none;
  color: ${({ theme }) => theme.palette.DARK_MAX};
  -webkit-tap-highlight-color: transparent;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
`;

export const AutocompleteMobile: React.FC<AutocompleteMobileProps> = ({
  suggestions,
  fetchSuggestions,
  selectSuggestion,
  loading,
  hasSearchIcon,
  className,
}) => {
  const listboxWrapperRef = useRef<HTMLDivElement>(null);
  const isSuggestable = useRef(true);
  const [value, setValue] = useState('');
  const [focusedPlaceholder, setFocusedPlaceholder] = useState(
    PlaceholderText.BLURRED,
  );
  const [labelTextShowed, setLabelTextShowed] = useState(
    PlaceholderText.BLURRED as string,
  );
  const [isModalShowed, setIsModalShowed] = useState(false);
  const [isListboxFocused, setIsListboxFocused] = useState(false);

  useEffect(() => {
    if (isModalShowed && listboxWrapperRef.current) {
      disableBodyScroll(listboxWrapperRef.current);
    } else if (!isModalShowed && listboxWrapperRef.current) {
      setValue('');
      setIsListboxFocused(false);
      enableBodyScroll(listboxWrapperRef.current);
    }
  }, [isModalShowed]);

  useEffect(() => {
    if (value.length > 2 && isSuggestable.current) {
      fetchSuggestions(value);
    }
  }, [value]);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const currentValue = event.target.value;

    setIsListboxFocused(currentValue.length > 2);
    setValue(currentValue);
  };

  const handleInputFocus = () => {
    if (!value) {
      setFocusedPlaceholder(PlaceholderText.FOCUSED);
    } else if (value.length > 2) {
      setIsListboxFocused(true);
    }
  };

  const handleInputBlur = () => {
    !value && setFocusedPlaceholder(PlaceholderText.BLURRED);
  };

  const handleSuggestionClick = (restaurantId: string, showedText: string) => {
    setValue(showedText);
    setLabelTextShowed(showedText);
    setIsModalShowed(false);
    isSuggestable.current = false;
    selectSuggestion(restaurantId, showedText);
  };

  const handleCloseModal = () => {
    setIsModalShowed(false);
    setIsListboxFocused(false);
  };

  return (
    <StyledAutocompleteMobile data-testid="autocomplete" className={className}>
      <>
        <StyledLabel data-testid="label">
          <StyledLabelButton
            type="button"
            onClick={() => setIsModalShowed(true)}
          >
            <i data-testid={'label-icon'} className="material-icons">
              search
            </i>
            <span>{labelTextShowed}</span>
          </StyledLabelButton>
        </StyledLabel>
        <StyledModal isShowed={isModalShowed} data-testid="modal">
          <StyleHeading>
            <StyleHeadingText text="Dublin restaurants" />
            <StyleHeadingButton type="button" onClick={handleCloseModal}>
              <i className="material-icons">close</i>
            </StyleHeadingButton>
          </StyleHeading>
          <StyledInput
            type="text"
            placeholder={focusedPlaceholder}
            onChange={handleChange}
            value={value}
            onFocus={handleInputFocus}
            onBlur={handleInputBlur}
            hasSearchIcon={hasSearchIcon}
          />
          <StyledListboxWrapper
            isShowed={isListboxFocused && suggestions.length > 0}
            data-testid="listbox-wrapper"
            ref={listboxWrapperRef}
          >
            {loading ? (
              <StyledLoader text={DEFAULT_TEXT_LOADING} />
            ) : (
              <StyledListbox role="listbox">
                {suggestions.map(
                  ({
                    id,
                    imgSrc,
                    title,
                    content,
                    route,
                    asRoute,
                  }: Restaurant) => (
                    <StyledListboxItem key={id} role="option">
                      <Link href={route} as={asRoute}>
                        <Card
                          imgSrc={imgSrc}
                          title={title}
                          content={content}
                          onClick={() => handleSuggestionClick(id, title)}
                          type={CardType.SUGGESTION}
                        />
                      </Link>
                    </StyledListboxItem>
                  ),
                )}
              </StyledListbox>
            )}
          </StyledListboxWrapper>
        </StyledModal>
      </>
    </StyledAutocompleteMobile>
  );
};
