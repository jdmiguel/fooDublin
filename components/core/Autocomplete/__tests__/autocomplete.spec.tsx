import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Autocomplete } from '../Autocomplete';

import { AUTOCOMPLETE_PROPS_MOCK } from '../__mocks__/autocomplete.mocks';

import { renderWithTheme } from '../../../../helpers/Theme';

describe('Component: Autocomplete', () => {
  it('should render', () => {
    const { container } = render(
      renderWithTheme(<Autocomplete {...AUTOCOMPLETE_PROPS_MOCK} />),
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not show suggestions list with a search of less than 3 characters', () => {
    const { getByPlaceholderText, getByTestId } = render(
      renderWithTheme(<Autocomplete {...AUTOCOMPLETE_PROPS_MOCK} />),
    );

    fireEvent.change(getByPlaceholderText('Search for locals...'), {
      target: { value: 'tr' },
    });
    expect(
      getByTestId('autocomplete-box-list').getAttribute('class'),
    ).not.toContain('active');
  });

  it('should show suggestions list with a search of 3 characters and call callback function', () => {
    const handleFetchSuggestion = jest.fn();
    const { getByPlaceholderText, getByTestId } = render(
      renderWithTheme(
        <Autocomplete
          {...AUTOCOMPLETE_PROPS_MOCK}
          fetchSuggestions={handleFetchSuggestion}
        />,
      ),
    );

    fireEvent.change(getByPlaceholderText('Search for locals...'), {
      target: { value: 'tre' },
    });
    expect(
      getByTestId('autocomplete-box-list').getAttribute('class'),
    ).toContain('active');
    expect(handleFetchSuggestion).toHaveBeenCalledTimes(1);
  });

  it('should show/hide suggestions list', async () => {
    const handleSelectSuggestion = jest.fn();
    const { getByPlaceholderText, getByTestId, findByPlaceholderText } = render(
      renderWithTheme(
        <Autocomplete
          {...AUTOCOMPLETE_PROPS_MOCK}
          selectSuggestion={handleSelectSuggestion}
        />,
      ),
    );
    const listbox = getByTestId('autocomplete-box-list');
    const input = getByPlaceholderText('Search for locals...');
    const closeButton = listbox.querySelector('button');

    // show options list by typing three characters
    fireEvent.change(input, { target: { value: 'tre' } });
    expect(listbox.getAttribute('class')).toContain('active');

    // hide options list by activating blur event
    fireEvent.blur(listbox);
    expect(listbox.getAttribute('class')).not.toContain('active');

    // show options list by activating focus event
    fireEvent.focus(input);
    expect(listbox.getAttribute('class')).toContain('active');

    // hide options list by clicking close buton
    fireEvent.click(closeButton);
    expect(listbox.getAttribute('class')).not.toContain('active');

    // show options list by activating focus event
    fireEvent.focus(input);
    expect(listbox.getAttribute('class')).toContain('active');

    // call callback function and hide options list by clicking any suggestion
    const firstSuggestionLink = listbox
      .querySelectorAll('li')[0]
      .querySelector('a');

    fireEvent.click(firstSuggestionLink);
    expect(handleSelectSuggestion).toHaveBeenCalledTimes(1);

    const inputUpdated = await findByPlaceholderText(
      'Start typing to search...',
    );
    const castedInput = inputUpdated as HTMLInputElement;

    expect(castedInput.value).toBe('Manifesto');
    expect(listbox.getAttribute('class')).not.toContain('active');
  });
});
