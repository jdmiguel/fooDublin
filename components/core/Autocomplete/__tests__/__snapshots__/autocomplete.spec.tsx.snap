// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Autocomplete should render 1`] = `
.c1 {
  width: 100%;
  border-radius: 4px;
  padding: 0 20px;
  background: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 i {
  font-size: 1.2rem;
  color: #A7A7A7;
  margin-right: 10px;
}

.c3 {
  width: 100%;
  height: 55px;
  padding: 10px 0;
  line-height: 20px;
  border: none;
  outline: none;
  color: #2f3235;
  background: transparent;
  cursor: pointer;
  outline: none;
  -webkit-tap-highlight-color: transparent;
}

.c7 {
  width: 100%;
  max-width: 600px;
  height: 100%;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 0 0 10px;
  cursor: pointer;
}

.c8 {
  width: 30px;
  height: 30px;
  border: solid #FFFFFF 1px;
  display: inline-block;
  border-radius: 4px;
  -webkit-appearance: button-bevel;
  box-shadow: 1px 2px 5px #888;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  padding: 5px 0;
  width: 65%;
  margin-left: 14px;
}

.c10 {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  width: 92%;
  font-size: 1.05rem;
  font-weight: 600;
  color: #FBA52B;
}

.c11 {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  width: 92%;
  font-size: 0.92rem;
  line-height: 1.1rem;
  font-weight: 500;
  color: #585858;
}

.c0 {
  width: 100%;
  max-width: 550px;
  height: 55px;
  position: relative;
  cursor: pointer;
}

.c2 {
  background-color: #F3F3F3;
  -webkit-transition: background-color 0.2s ease-out;
  transition: background-color 0.2s ease-out;
}

.c2:hover {
  background-color: #FDF3E5;
}

.c4 {
  visibility: hidden;
  opacity: 0;
  -webkit-transform: translateY(10px);
  -ms-transform: translateY(10px);
  transform: translateY(10px);
  -webkit-transition: opacity 0.2s ease 0s,-webkit-transform 0.2s ease 0s;
  -webkit-transition: opacity 0.2s ease 0s,transform 0.2s ease 0s;
  transition: opacity 0.2s ease 0s,transform 0.2s ease 0s;
  position: absolute;
  box-sizing: border-box;
  background: #F3F3F3;
  z-index: 2;
  top: 55px;
  left: 0;
  padding: 0;
  width: 100%;
  height: auto;
  max-height: 400px;
  overflow: auto;
  box-shadow: 0 2px 2px rgba(0,0,0,0.35);
  border: 1px solid #F3F3F3;
  border-radius: 0 0 4px 4px;
  border-bottom-right-radius: 4px;
  border-bottom-left-radius: 4px;
  border-top: 0;
}

.c5 {
  width: 100%;
  -webkit-tap-highlight-color: transparent;
}

.c6 {
  width: 100%;
  border-bottom: 1px solid #E5EAED;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #F3F3F3;
  -webkit-transition: background-color 0.2s ease-out;
  transition: background-color 0.2s ease-out;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  cursor: pointer;
}

.c6:hover {
  background-color: #FDF3E5;
}

@media only screen and (min-width:768px) {
  .c1 {
    max-width: 550px;
  }
}

@media only screen and (min-width:768px) {
  .c7 {
    -webkit-transition: background-color 0.2s ease-out;
    transition: background-color 0.2s ease-out;
  }

  .c7:hover {
    background-color: #FDF3E5;
  }

  .c7:hover h4 {
    color: #EC7E00;
  }
}

@media only screen and (min-width:768px) {
  .c9 {
    font-size: 1.3rem;
    line-height: 1.3rem;
    padding: 10px 0;
    margin-left: 16px;
  }
}

@media only screen and (min-width:1200px) {
  .c9 {
    margin-left: 20px;
  }
}

@media only screen and (min-width:768px) {
  .c10 {
    -webkit-transition: color 0.2s ease-out;
    transition: color 0.2s ease-out;
  }
}

<div
  class="c0"
  data-testid="autocomplete"
>
  <div
    class="c1 c2"
    data-testid="input-wrapper"
  >
    <i
      class="material-icons"
      data-testid="input-icon"
    >
      search
    </i>
    <input
      class="c3"
      placeholder="Search for locals..."
      type="text"
      value=""
    />
  </div>
  <div
    class="c4"
    data-testid="listbox-wrapper"
  >
    <ul
      class="c5"
      role="listbox"
    >
      <li
        class="c6"
        role="option"
      >
        <a
          class="c7"
          data-testid="card"
          type="suggestion"
        >
          <div
            class="LazyImage-Placeholder"
          >
            <img
              alt="Manifesto"
              class="c8"
              src="http://localhost:3000/static/images/generic-thumb.png"
              type="suggestion"
            />
          </div>
          <div
            class="c9"
            type="suggestion"
          >
            <h4
              class="c10"
              type="suggestion"
            >
              Manifesto
            </h4>
            <p
              class="c11"
              type="suggestion"
            >
              Rathmines
            </p>
          </div>
        </a>
      </li>
      <li
        class="c6"
        role="option"
      >
        <a
          class="c7"
          data-testid="card"
          type="suggestion"
        >
          <div
            class="LazyImage-Placeholder"
          >
            <img
              alt="Dragon Inn"
              class="c8"
              src="http://localhost:3000/static/images/generic-thumb.png"
              type="suggestion"
            />
          </div>
          <div
            class="c9"
            type="suggestion"
          >
            <h4
              class="c10"
              type="suggestion"
            >
              Dragon Inn
            </h4>
            <p
              class="c11"
              type="suggestion"
            >
              South City West
            </p>
          </div>
        </a>
      </li>
      <li
        class="c6"
        role="option"
      >
        <a
          class="c7"
          data-testid="card"
          type="suggestion"
        >
          <div
            class="LazyImage-Placeholder"
          >
            <img
              alt="Casa Del Toro"
              class="c8"
              src="http://localhost:3000/static/images/generic-thumb.png"
              type="suggestion"
            />
          </div>
          <div
            class="c9"
            type="suggestion"
          >
            <h4
              class="c10"
              type="suggestion"
            >
              Casa Del Toro
            </h4>
            <p
              class="c11"
              type="suggestion"
            >
              Drumcondra
            </p>
          </div>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
